### PR TITLE
Feature/move alerts

### PR DESCRIPTION
**Pull request by sandypadmanabhan**
### What
To refactor slack Alarm channel and slack warning channel to use the new library 

Trello:
https://trello.com/c/2VlcYQ4u/425-move-all-alerts-currently-going-to-system-messages-s-m
PR created for config changes
In Develop- 'https://github.com/ONSdigital/dp-configs/pull/355'
In Production- 'https://github.com/ONSdigital/dp-configs/pull/354'

### How to review

Check for typo
Check if changes makes sense

### Who can review

Anyone

### CODE REVIEW

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/model/approval/tasks/timeseries/TimeSeriesCompressionTask.java [hunk 1]
```diff
@@ -88,11 +90,12 @@ public boolean compressTimeseries(Collection collection, CollectionReader collec
...
-                    "Failed verification of time series zip file",
-                    new PostMessageField("Attempt", Integer.toString(attempt, 10), true),
-                    new PostMessageField("Zip path", failedZipFile.zipPath.toString(), false)
-            );
+
+            String channel = Root.zebedee.getSlackCollectionAlarmChannel();
+            AttachmentField attemptField = new AttachmentField("Attempt", Integer.toString(attempt, 10), true);
+            AttachmentField zipPath= new AttachmentField("Zip path", failedZipFile.zipPath.toString(), false);
+
+            Root.zebedee.getSlackNotifier().callCollectionWarning(collection, channel, "Failed verification of time series zip file",attemptField, zipPath);
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/model/approval/tasks/timeseries/TimeSeriesCompressionTask.java:98]: We can move the channel and the get notifier calls outside of the loop to make it more efficient:

```
String channel = Root.zebedee.getSlackCollectionAlarmChannel();
Notifier notifier =  Root.zebedee.getSlackNotifier();

for (TimeseriesCompressionResult failedZipFile : failedZipFiles) {
    AttachmentField attemptField = new AttachmentField("Attempt", Integer.toString(attempt, 10), true);
    AttachmentField zipPath= new AttachmentField("Zip path", failedZipFile.zipPath.toString(), false);
    
notifier.callCollectionWarning(collection, channel, "Failed verification of time series zip file",attemptField, zipPath);
}
```

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/AttachmentField.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package com.github.onsdigital.zebedee.util.slack;
+
+public class AttachmentField {
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/AttachmentField.java:3]: Add javadoc to explain purpose of class.

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/AttachmentField.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    private String title;
+    private String message;
+    private boolean isShort;
+
+    public AttachmentField(String title, String message, boolean isShort) {
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/AttachmentField.java:8]: Add javadoc for constructor.

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/Notifier.java [hunk 4]
```diff
@@ -17,12 +19,23 @@
...
+
+    void sendSlackMessage(PostMessage message) throws Exception;
+
+    PostMessage createPostMessage(String channel, String text);
+
+    void callCollectionAlarm(Collection c, String channel, String message, Exception ex);
+
+    void callCollectionWarning(Collection c, String channel, String message, AttachmentField... attachments);
+
+    void callCollectionAlarm(Collection c, String channel, String message, AttachmentField... attachments);
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/Notifier.java:38]: Need to write Javadoc for each of these methods to describe what they do.

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/Notifier.java [hunk 5]
```diff
@@ -17,12 +19,23 @@
...
      * @param alarm - the string message to apply to the notification.
      * @param args - additional arguments to add to the notification.
      */
     void alarm(String alarm, PostMessageField... args);
+
+    void sendSlackMessage(PostMessage message) throws Exception;
+
+    PostMessage createPostMessage(String channel, String text);
+
+    void callCollectionAlarm(Collection c, String channel, String message, Exception ex);
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/Notifier.java:34]: I think we should rename these 3 methods - `sendCollectionAlarm`/ `sendCollectionWarning` sounds more appropriate to me.

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/SlackNotifier.java [hunk 6]
```diff
@@ -32,4 +50,70 @@ public void collectionAlarm(Collection c, String alarm, PostMessageField... args
...
+                .addField("CollectionID", collection.getId(), false)
+                .addField("Collection Name", collection.getDescription().getName(), false);
+
+        if (fields != null)  {
+            for(AttachmentField field : fields) {
+                attachment.addField(field.getTitle(), field.getMessage(), field.isShort());
+            }
+        }
+
+        return attachment;
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/SlackNotifier.java:116]: You also need to add tests for the new functionality you've added to this class.

**Review by daiLlew — CHANGES_REQUESTED** [review]

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/ZebedeeConfiguration.java [hunk 7]
```diff
@@ -9,12 +9,7 @@
 import com.github.onsdigital.slack.client.SlackClient;
 import com.github.onsdigital.slack.client.SlackClientImpl;
 import com.github.onsdigital.zebedee.data.processing.DataIndex;
-import com.github.onsdigital.zebedee.keyring.CentralKeyringImpl;
-import com.github.onsdigital.zebedee.keyring.Keyring;
-import com.github.onsdigital.zebedee.keyring.KeyringException;
-import com.github.onsdigital.zebedee.keyring.KeyringMigratorImpl;
-import com.github.onsdigital.zebedee.keyring.LegacyKeyringImpl;
-import com.github.onsdigital.zebedee.keyring.NoOpCentralKeyring;
+import com.github.onsdigital.zebedee.keyring.*;
```
[the remark this answers is not in the document]
Comment (sandypadmanabhan) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/ZebedeeConfiguration.java:12]: I tried removing the wildcard and importing just the required one but it always goes back to the wild card

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/configuration/Configuration.java [hunk 8]
```diff
@@ -20,6 +20,7 @@
 public class Configuration {
 
     private static final String DEFAULT_WEBSITE_URL = "http://localhost:8080";
+    private static final String DEFAULT_SLACK_CHANNEL= "slack-client-test";
```
Comment (sandypadmanabhan) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/configuration/Configuration.java:23]: Not sure if I should change the value of this variable?

File: zebedee-cms/src/test/java/com/github/onsdigital/zebedee/util/slack/SlackNotifierTest.java [hunk 9]
```diff
@@ -0,0 +1,576 @@
...
+
+    @Test
+    public void sendCollectionAlarmChannelEmpty() {
+        SlackNotifier sn = new SlackNotifier(slackClient);
+        boolean returnValue = sn.sendCollectionAlarm(collection, "", "message", new Exception("Hello test"));
+
+        verify(slackClient, times(0)).sendMessage(pm);
+        assertThat(pm.getAttachments().size(), equalTo(0));
+        assertFalse(returnValue);
+    }
```
Comment (daiLlew) [inline zebedee-cms/src/test/java/com/github/onsdigital/zebedee/util/slack/SlackNotifierTest.java:116]: `channel` can also be `null`. Duplicate this test but pass in `null` rather than `""`.

I know this feels possibly over the top _but_ it is possible that `null` could be provided and that will affect how this code behaves therefore we should confirm the behaviour with a test to prove it works as we expect.

Same for `message` and `exception` - add test cases to prove it behaves as expected if they are null.

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/Zebedee.java [hunk 10]
```diff
@@ -426,4 +437,21 @@ public EncryptionKeyFactory getEncryptionKeyFactory() {
...
+        return this.slackClient;
+    }
+
+    public String getSlackCollectionAlarmChannel(){
+        return this.slackCollectionAlarmChannel;
+    }
+
+    public String getSlackCollectionWarningChannel(){
+        return this.slackCollectionWarningChannel;
+    }
```
[the remark this answers is not in the document]
Comment (sandypadmanabhan) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/Zebedee.java:451]: If this is the case, we need t add 2 config variables in the configuration class, one for warning channel and one for alarm channel. Currently both will grab information from the same env. variable. Are we ok with that Dave?

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/SlackNotifier.java [hunk 11]
```diff
@@ -26,10 +42,112 @@ public void collectionAlarm(Collection c, String alarm, PostMessageField... args
...
+        if (StringUtils.isEmpty(channel) || StringUtils.isEmpty(customMessage) || collection == null ) {
+            error().log("error sending collection alarm. channel/customMessage/collection was empty");
+            return false;
+        }
+        if ( collection.getDescription() == null || collection.getDescription().getType() == null
+                || StringUtils.isEmpty(collection.getId()) || StringUtils.isEmpty(collection.getDescription().getName()))
+        {
+            error().log("error sending collection alarm. Collection description/id/type/name is null");
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (sandypadmanabhan) [reply zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/SlackNotifier.java:103]: Ah , yes. Sorry I missed that point.

File: zebedee-cms/src/test/java/com/github/onsdigital/zebedee/util/slack/SlackNotifierTest.java (hunk 9 above)
Comment (sandypadmanabhan) [reply zebedee-cms/src/test/java/com/github/onsdigital/zebedee/util/slack/SlackNotifierTest.java:116]: I have added the test check ex.getMessage() to be null or empty. Not sure, if I should be checking the exception itself rather than ex.getMessage()

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/SlackNotifier.java [hunk 12]
```diff
@@ -26,10 +42,116 @@ public void collectionAlarm(Collection c, String alarm, PostMessageField... args
...
+    public PostMessage createPostMessage(String channel, String text) {
+        return profile.newPostMessage(channel, text);
+    }
+
+    @Override
+    public boolean sendCollectionAlarm(Collection collection, String channel, String customMessage, Exception ex) {
+        boolean result = validate(collection, channel, customMessage, "alarm");
+        if (!result) {
+            return false;
+        }
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/SlackNotifier.java:72]:
```
if !validate(collection, channel, customMessage, "alarm") {
    return false;
}
```
Functionally the same just shorter 😄 . Generally speaking I would only assign the result to a variable if I need to access it multiple times. It's also good to be consistent in how you return from a method:
Either do:
```
if !validate(collection, channel, customMessage, "alarm") {
    return false;
}
```
or
```
boolean isValid = validate(collection, channel, customMessage, "alarm");
if (!isValid) {
    return false;
}
```
But don't mix and match. I know it seems like a small point but you should try either follow the existing style thats in place or choose your own approach and implement it consistently - this helps to keep the code readable.

I am being really picky with this comment, the rest of this code is looking good so if you address this comment we can merge this in.

File: zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/SlackNotifier.java [hunk 13]
```diff
@@ -26,10 +42,116 @@ public void collectionAlarm(Collection c, String alarm, PostMessageField... args
...
+        AttachmentField exField = new AttachmentField("exception", ex.getMessage(), false);
+        PostMessage postMessage = createPostMessage(channel, customMessage)
+                .addAttachment(createCollectionAttachment("Alert", "Collection Alarm", Colour.DANGER, collection, exField));
+
+        boolean messageOutput = postMessageToSlack(postMessage);
+        if (!messageOutput) {
+            return false;
+        }
```
Comment (daiLlew) [inline zebedee-cms/src/main/java/com/github/onsdigital/zebedee/util/slack/SlackNotifier.java:83]:
```
if (!postMessageToSlack(postMessage)) {
            return false;
}
```

**Review by daiLlew — APPROVED** [review]
